packages/core/src/api/serialization/serialization.ts

```diff
@@ -0,0 +1,182 @@
+        const blockContent = DOMSerializer.renderSpec(
+          doc(options),
+          editor.schema[node.firstChild!.type.name as keyof BSchema].serialize!(
+            nodeToBlock(
+              node,
+              editor.schema,
+              editor.blockCache
+            ) as SpecificBlock<BlockSchema, string>,
+            editor as BlockNoteEditor<BlockSchema>
+          )
+        );
+
+        // Renders inline content.
+        if (blockContent.contentDOM) {
+          if (node.isLeaf) {
+            throw new RangeError(
+              "Content hole not allowed in a leaf node spec"
+            );
+          }
+
+          blockContent.contentDOM.appendChild(
+            customSerializer.serializeFragment(
+              node.firstChild!.content,
+              options
+            )
+          );
+        }
+
+        contentDOM.appendChild(blockContent.dom);
+
+        // Renders nested blocks.
+        if (node.childCount === 2) {
+          customSerializer.serializeFragment(
+            Fragment.from(node.content.lastChild),
+            options,
+            contentDOM
+          );
+        }
+      } else {
+        // Renders the block normally, i.e. using `toDOM`.
+        customSerializer.serializeFragment(node.content, options, contentDOM);
+      }
+    }
+
+    return dom as HTMLElement;
+  };
+
+  return customSerializer;
+};
+
+export const createCustomBlockSerializerExtension = <
+  BSchema extends BlockSchema
+>(
+  editor: BlockNoteEditor<BSchema>
+) =>
+  Extension.create<{ editor: BlockNoteEditor<BSchema> }, undefined>({
+    addProseMirrorPlugins() {
+      const tiptap = this.editor;
+      const schema = this.editor.schema;
+      return [
+        new Plugin({
+          props: {
+            // TODO: Totally broken on Firefox as it outright doesn't allow
+            //  reading from or writing to the clipboard.
+            handleDOMEvents: {
+              copy(_view, event) {
+                // Stops the default browser copy behaviour.
+                event.preventDefault();
+                event.clipboardData!.clearData();
+
+                async function setClipboardData() {
+                  const serializer = customBlockSerializer(schema, editor);
+
+                  const selectedFragment =
+                    tiptap.state.selection.content().content;
+
+                  const selectedHTML =
+                    serializer.serializeFragment(selectedFragment);
+
+                  const parent = document.createElement("div");
+                  parent.appendChild(selectedHTML);
+
+                  const structured = parent.innerHTML;
+                  const clean = await cleanHTML(structured);
+                  const plain = await markdown(clean);
+
+                  event.clipboardData!.setData("text/plain", plain);
+                  event.clipboardData!.setData("text/html", clean);
+                  // TODO: Writing to other MIME types not working in Safari for
```

Comment:
Ah good point, in any case that's still a problem since the HTML cleaning itself is async

Reply:
true, but first let's see if this was the issue or not

Reply:
This was indeed the issue for Firefox. Safari, like Chrome, doesn't seem to care if the clipboard contents are set asynchronously, regardless it won't write to `blocknote/html`.

Reply:
Judging from notion web, when using on safari it also doesn't add custom data to the clipboard. (only sanitized html).

I'm not sure yet how they do it, because when I copy and paste within notion on safari (even across tabs or between safari private mode and regular), colored text works (even though that info is not on the clipboard)